Update to LoginState.java. If oldSession is null do not upgrade the session and create a new one.

pmolchanov2002 opened the pull request:
If oldSession is null or oldUserDN is null do not upgrade the session and create a new one.

If user creates 2 different sessions within a few milliseconds, a race condition between threads may occur that can cause a "Session Upgrade fails since user is different than the original authenticated user" exception. oldSession may be null at the point of the session upgrade. If oldSession is null, session upgrade doesn't make sense and sessionUpgrade flag should be set to false to avoid an exception.

pmolchanov2002 committed:
Update to LoginState.java. If oldSession is null or oldUserDN is null, do not upgrade the session and create a new one.

If user creates 2 different sessions within few milliseconds, a race condition between threads may occur that can cause a "Session Upgrade fails since user is different than original authenticated user" exception. oldSession may be null at the point of the session upgrade. If oldSession is null, session upgrade doesn't make sense and sessionUpgrade flag should be set to false to avoid an exception.

maximthomas commented:
@pmolchanov2002 Hi, default value for sessionUpgrade is false, so it's true value it has been set earlier. I think `isSessionUpgrade` should be initialised properly earlier in the callstack in `AuthUtils.getAuthContextMethod`

https://github.com/OpenIdentityPlatform/OpenAM/blob/733caa9fe8bdbf72492390530053d311cdeff8ea/openam-core/src/main/java/com/sun/identity/authentication/service/AuthUtils.java#L899-L905

pmolchanov2002 commented:
As you may see from the logs, the getOldSession is null even if the session upgrade is true. How can it be null at this point?

```
activate before isSessionUpgrade=true forceAuth=false getSessionActivator=DefaultSessionActivator oldSessionReference=AQIC5wM2LY4SfcxkR8QK_mItMpLRpGbUHy0S_WP9DCXrP
0o.AAJTSQACMDIAAlNLABM1MjM1ODI5MjU4ODk4Nzg4NTc4AAJTMQACMDQ. getOldSession=null sessionReference=AQIC5wM2LY4SfcwR2XNgpV5SIAo2fwtaTCXd2Zn4LuplFcw.AAJTSQACMDIAAlNL
ABQtMzc1NzgwNzI5MjM2MDYwMDcyNwACUzEAAjA0
```

pmolchanov2002 commented:
Thank you, Maxim. Good point!

pmolchanov2002 commented:
Hi,

I think this code should be changed in the OpenAM/openam-core/src/main/java/com/sun/identity/authentication/service/AuthUtils.java at line 993:

```
                    InternalSession oldSession = null;
                    if (sid != null) {
                        oldSession = AuthD.getSession(sid);
                        loginState.setOldSession(oldSession);
                    }
                    if (sessionUpgrade) {
                        loginState.setOldSession(oldSession);
                        loginState.setSessionUpgrade(sessionUpgrade);
                    }
```

to

```
                        loginState.setOldSession(sess);
                        loginState.setSessionUpgrade(sessionUpgrade);
```


The old session was found and verified way before at this point in the `sess` variable:

```
                sid = new SessionID(sessionID);
                authContext = retrieveAuthContext(sid);
                
                // check if this sesson id is active, if yes then it
                // is a session upgrade case.
                loginState = getLoginState(authContext);
                if (loginState != null) {
                    sess = loginState.getSession();
                } else {
                    sess = AuthD.getSession(sessionID);
                }
                if (sess == null) {
                    sessionUpgrade = false;
                }
```

pmolchanov2002 commented:
We can add an extra check for the old session as well:

                sid = new SessionID(sessionID);
                authContext = retrieveAuthContext(sid);
                
                // check if this sesson id is active, if yes then it
                // is a session upgrade case.
                loginState = getLoginState(authContext);
                if (loginState != null) {
                    sess = loginState.getSession();
                } else {
                    sess = AuthD.getSession(sessionID);
                }
                // Session upgrade can be done only if old session is not null
                if (sess == null || AuthD.getSession(sessionID) == null) {
                    sessionUpgrade = false;
                }

pmolchanov2002 committed:
Update LoginState.java

pmolchanov2002 committed:
Update AuthUtils.java

Added a null check for the old session before assignment to the LoginState and assigned old session to previously identified session

pmolchanov2002 committed:
Update AuthUtils.java

Corrected assignments of the oldSession

maximthomas commented:
@pmolchanov2002 Hi, seems ok.
Please remove added commented lines in LoginState.java file

pmolchanov2002 committed:
Update LoginState.java

Removed commented code for the previous fix attempt

pmolchanov2002 commented:
Maxim, thank you for the review and your suggestions! This was a pleasant experience working with you on this issue.

maximthomas commented:
@pmolchanov2002, glad to know that, thank you for the pull request!
